sycl/source/detail/graph_impl.cpp

```diff
@@ -1091,6 +1149,252 @@ void exec_graph_impl::duplicateNodes() {
+  }
+
+  // If there are any accessor requirements, we have to update through the
+  // scheduler to ensure that any allocations have taken place before trying to
+  // update.
+  bool NeedScheduledUpdate = false;
+  for (auto &Node : Nodes) {
+    // Check if node(s) derived from this modifiable node exists in this graph
+    if (MIDCache.count(Node->getID()) == 0) {
+      throw sycl::exception(
+          sycl::make_error_code(errc::invalid),
+          "Node passed to update() is not part of the graph.");
+    }
+    if (Node->MCGType != sycl::detail::CG::Kernel) {
+      throw sycl::exception(errc::invalid, "Cannot update non-kernel nodes");
+    }
+
+    if (Node->MCommandGroup->getRequirements().size() == 0) {
+      continue;
+    }
+    NeedScheduledUpdate = true;
+    // Update cached requirements for this graph with updated node ones
+    auto UpdatedReqs = Node->MUpdatedAccessorsCache;
+    for (auto &CachedReq : MRequirements) {
+      for (auto &UpdatedReq : UpdatedReqs) {
+        if (CachedReq == UpdatedReq.first) {
+          CachedReq = UpdatedReq.second;
+        }
+      }
+    }
+  }
+
+  // Clean up any execution events which have finished so we don't pass them to
+  // the scheduler.
+  for (auto It = MExecutionEvents.begin(); It != MExecutionEvents.end();) {
+    if ((*It)->isCompleted()) {
+      It = MExecutionEvents.erase(It);
+      continue;
+    }
+    ++It;
+  }
+
+  // If we have previous execution events do the update through the scheduler to
+  // ensure it is ordered correctly.
+  NeedScheduledUpdate |= MExecutionEvents.size() > 0;
+
+  if (NeedScheduledUpdate) {
+    auto AllocaQueue = std::make_shared<sycl::detail::queue_impl>(
+        sycl::detail::getSyclObjImpl(MGraphImpl->getDevice()),
+        sycl::detail::getSyclObjImpl(MGraphImpl->getContext()),
+        sycl::async_handler{}, sycl::property_list{});
+    // Don't need to care about the return event here because it is synchronous
+    sycl::detail::Scheduler::getInstance().addCommandGraphUpdate(
+        this, Nodes, AllocaQueue, MRequirements, MExecutionEvents);
+  } else {
+    for (auto &Node : Nodes) {
+      updateImpl(Node);
+    }
+  }
+}
+
+void exec_graph_impl::updateImpl(std::shared_ptr<node_impl> Node) {
+  auto ContextImpl = sycl::detail::getSyclObjImpl(MContext);
+  const sycl::detail::PluginPtr &Plugin = ContextImpl->getPlugin();
+  auto DeviceImpl = sycl::detail::getSyclObjImpl(MGraphImpl->getDevice());
+
+  // Gather arg information from Node
+  auto &ExecCG =
+      *(static_cast<sycl::detail::CGExecKernel *>(Node->MCommandGroup.get()));
+  // Copy args because we may modify them
+  std::vector<sycl::detail::ArgDesc> NodeArgs = ExecCG.getArguments();
+  // Copy NDR desc since we need to modify it
+  auto NDRDesc = ExecCG.MNDRDesc;
+
+  // TODO: Try and refactor this stuff because it is copied from
+  // enqueueImpCommandBufferKernel
+  // TODO: Probably have to handle eliminated args too
```

Comment:
These TODOs intended for this PR? If not can you create an internal ticket so we don't forget this improvement